### PR TITLE
Update mypy to 1.7.0

### PR DESCRIPTION
**Pull request by int3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114160
* #114162

It appears that `mypy` is now checking a few more previously-unchecked files; these files
are being found via import-following. Not sure exactly why they weren't being checked before.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @mrshenli @zhaojuanmao @rohan-varma @kiukchung @d4l3k @lucasllc @XilunWu @tianyu-l

### CODE REVIEW

File: mypy-inductor.ini [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [mypy]
 plugins = mypy_plugins/check_mypy_version.py, numpy.typing.mypy_plugin
 
-cache_dir = .mypy_cache/nofollow
+cache_dir = .mypy_cache/inductor
```
Comment (int3) [inline mypy-inductor.ini:4]: stray change that should have been part of https://github.com/pytorch/pytorch/pull/113830

File: torch/autograd/profiler_util.py [hunk 2]
```diff
@@ -774,11 +774,11 @@ class MemRecordsAcc:
...
     def __init__(self, mem_records):
         self._mem_records = mem_records
-        self._start_uses = []
-        self._indices = []
+        self._start_uses: List[int] = []
+        self._indices: List[int] = []
         if len(mem_records) > 0:
             tmp = sorted([(r[0].start_us(), i) for i, r in enumerate(mem_records)])
-            self._start_uses, self._indices = zip(*tmp)
+            self._start_uses, self._indices = zip(*tmp)  # type: ignore[assignment]
```
Comment (int3) [inline torch/autograd/profiler_util.py:781]: `zip(*tmp)` returns a list of two tuples but `mypy` is unable to see that

File: torch/distributed/fsdp/_optim_utils.py [hunk 3]
```diff
@@ -340,7 +340,7 @@ def _broadcast_processed_state(
     if fsdp_state.rank == 0:
         objects[0] = tree_map_only(
             torch.Tensor,
-            lambda v: v.cpu() if v.dim() == 0 else _PosDimTensorInfo(v.shape, v.dtype),
+            lambda v: v.cpu() if v.dim() == 0 else _PosDimTensorInfo(v.shape, v.dtype),  # type: ignore[union-attr]
```
Comment (int3) [inline torch/distributed/fsdp/_optim_utils.py:343]: this seems like a `mypy` bug; somehow it thinks `v` has a `Never` type (afaik that is just the uninhabited type.)

**Comment by int3** [comment]
hm, CI seems to be having issues installing the new package

**Review by eellison — APPROVED** [review]

**Pull request closed without merging**